crates/ruff_python_parser/src/parser/statement.rs

```diff
@@ -100,69 +100,96 @@ impl<'src> Parser<'src> {
         ) || token.is_keyword()
     }
 
-    /// Parses a compound or a simple statement.
+    /// Parses a compound or a single simple statement.
+    ///
+    /// See:
+    /// - <https://docs.python.org/3/reference/compound_stmts.html>
+    /// - <https://docs.python.org/3/reference/simple_stmts.html>
     pub(super) fn parse_statement(&mut self) -> Stmt {
-        let start_offset = self.node_start();
+        let start = self.node_start();
+
         match self.current_token_kind() {
             TokenKind::If => Stmt::If(self.parse_if_statement()),
-            TokenKind::For => Stmt::For(self.parse_for_statement(start_offset)),
+            TokenKind::For => Stmt::For(self.parse_for_statement(start)),
             TokenKind::While => Stmt::While(self.parse_while_statement()),
-            TokenKind::Def => {
-                Stmt::FunctionDef(self.parse_function_definition(vec![], start_offset))
-            }
-            TokenKind::Class => Stmt::ClassDef(self.parse_class_definition(vec![], start_offset)),
+            TokenKind::Def => Stmt::FunctionDef(self.parse_function_definition(vec![], start)),
+            TokenKind::Class => Stmt::ClassDef(self.parse_class_definition(vec![], start)),
             TokenKind::Try => Stmt::Try(self.parse_try_statement()),
-            TokenKind::With => Stmt::With(self.parse_with_statement(start_offset)),
+            TokenKind::With => Stmt::With(self.parse_with_statement(start)),
             TokenKind::At => self.parse_decorators(),
             TokenKind::Async => self.parse_async_statement(),
             TokenKind::Match => Stmt::Match(self.parse_match_statement()),
             _ => self.parse_single_simple_statement(),
         }
     }
 
-    /// Parses a single simple statement, expecting it to be terminated by a newline or semicolon.
-    /// TODO(micha): It's not entirely clear why this method is necessary. It is called from
-    /// `parse_body` and it only reads out the first simple statement before calling `parse_statement` again.
-    /// This makes me wonder if the parser incorrectly allows `a;if b: pass`
```

Comment:
The reason this is required is because there's no dedicated node to capture a list of simple statements on the same line. This means that each statement needs to be parsed and returned one by one. This requires having a dedicated function to only parse a single statement.

Pyright has a dedicated node for this (`StatementListNode`) but this clutters the AST because then every single simple statement will be a `StatementListNode` with a single simple statement in it. [Pyright AST playground](https://zzzen.github.io/pyright-ast-viewer/#code/IYbgBARuDGBQywrOBLAZmAKgJwK4FMAuWMMUSEsyiIA)